File: app/src/main/java/im/fdx/v2ex/network/cookie/MyCookieJar.java
package im.fdx.v2ex.network.cookie;

import android.util.DebugUtils;

import com.elvishew.xlog.XLog;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.HttpUrl;

/**
 * Created by fdx on 2017/3/16.
 * <p>
 * 请用自己的Cookie策略，而不是开源库。
 */
public class MyCookieJar implements CookieJar {
    private final HashMap<String, List<Cookie>> cookieStore = new HashMap<>();
    private List<Cookie> mCookies = new ArrayList<>();
    private CookiePersistor cookiePersistor;

    public MyCookieJar(CookiePersistor persistor) {
        this.cookiePersistor = persistor;

        mCookies.addAll(cookiePersistor.loadAll());

    }

    @Override
    public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
        mCookies.clear();
        mCookies.addAll(cookies);

        cookiePersistor.persistAll(cookies);
    }

    @Override
    public List<Cookie> loadForRequest(HttpUrl url) {


        List<Cookie> cookies;
        if (cookiePersistor.loadAll() != null) {
            cookies = cookiePersistor.loadAll();
        } else
            cookies = cookieStore.get(url.host());


        XLog.e(cookies);

        return cookies == null ? new ArrayList<Cookie>() : cookies;
    }


    public void clear() {

        mCookies.clear();
        cookiePersistor.clear();

    }
}

File: app/src/main/java/im/fdx/v2ex/network/cookie/MyCookieJar.kt
package im.fdx.v2ex.network.cookie

import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import java.util.*

/**
 * Created by fdx on 2017/3/16.
 *
 *
 * 请用自己的Cookie策略，而不是开源库。
 */
class MyCookieJar(private val cookiePersistor: SharedPrefsPersistor) : CookieJar {
    private val cookieStore = HashMap<String, List<Cookie>>()
    private val mCookies = ArrayList<Cookie>()

    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        mCookies.clear()
        mCookies.addAll(cookies)
        cookieStore[url.host] = mCookies
        cookiePersistor.persistAll(cookies)
    }

    override fun loadForRequest(url: HttpUrl): List<Cookie> {

        val cookies: List<Cookie>? = if (cookiePersistor.loadByHost(url.host).isNotEmpty()) {
            cookiePersistor.loadByHost(url.host.removePrefix("www."))
        } else
            cookieStore[url.host]
        return cookies ?: ArrayList()
    }

    fun clear() {
        mCookies.clear()
        cookieStore.clear()
        cookiePersistor.clear()
    }
}
